realm/realm-library/src/objectServer/java/io/realm/internal/async/RealmAsyncResultTaskImpl.java

```diff
@@ -0,0 +1,142 @@
+
+    private Future<?> pendingTask;
+    private volatile boolean isCancelled = false;
+    private final ThreadPoolExecutor service;
+    private RealmNotifier handler = new AndroidRealmNotifier(null, new AndroidCapabilities());
+    private Executor<T> executor;
+
+    /**
+     * FIXME
+     *
+     * @param service
+     * @param executor
+     */
+    public RealmAsyncResultTaskImpl(ThreadPoolExecutor service, Executor<T> executor) {
+        this.service = service;
+        this.executor = executor;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void cancel() {
+        if (pendingTask != null) {
+            pendingTask.cancel(true);
+            isCancelled = true;
+
+            // From "Java Threads": By Scott Oaks & Henry Wong
+            // cancelled tasks are never executed, but may
+            // accumulate in work queues, which may causes a memory leak
+            // if the task hold references (to an enclosing class for example)
+            // we can use purge() but one caveat applies: if a second thread attempts to add
+            // something to the pool (using the execute() method) at the same time the
+            // first thread is attempting to purge the queue the attempt to purge
+            // the queue fails and the cancelled object remain in the queue.
+            // A better way to cancel objects with thread pools is to use the remove()
+            service.getQueue().remove(pendingTask);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean isCancelled() {
+        return isCancelled;
+    }
+
+    @Override
+    public T blockingGet() {
+        return executor.run();
+    }
+
+    @Override
+    public void get(App.Callback<T> callback) {
+        pendingTask = service.submit(new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    postSuccess(executor.run(), callback);
+                } catch (AppException e) {
+                    postError(e, callback);
+                } catch (Throwable e) {
+                    postError(new AppException(ErrorCode.UNKNOWN, "Unexpected error", e), callback);
+                }
+            }
+        });
+    }
+
+    private void postError(final AppException error, App.Callback<T> callback) {
+        boolean errorHandled = false;
+        if (callback != null) {
+            Runnable action = new Runnable() {
+                @Override
+                public void run() {
+                    callback.onResult(App.Result.withError(error));
+                }
+            };
+            errorHandled = handler.post(action);
+        }
+
+        if (!errorHandled) {
+            RealmLog.error(error, "An error was thrown, but could not be posted: \n" + error.toString());
+        }
+    }
+
+    private void postSuccess(final T result, App.Callback<T> callback) {
+        if (callback != null) {
+            handler.post(new Runnable() {
```

Comment:
There isn't anything checking if you call this on a thread without a Looper. Not exactly sure what happens, but it probably blows up with some weird exception. There also seem to tests lacking around this.

Ideally, we should throw an exception when you call `task.get(callback)` if the caller thread doesn't have a Looper.

Reply:
I wasn't totally on the right page in regards to this, but it doesn't seem to be crashing. The `AndroidRealmNotifier.post` method makes this check: `handler != null && handler.post(runnable)`, and we only have a `handler` if we have a looper, i.e.:
```
if (capabilities.canDeliverNotification()) {
            handler = new Handler(Looper.myLooper());
        } else {
            handler = null;
        }
```
It looks like `post` will simply return false - I don't know if this is expected behaviour though?

I will add a call to `Util.checkLooperThread` from the task class nonetheless.